File: 算法/代码随想录算法PDF/00调试源代码/968.监控二叉树.py
#
# @lc app=leetcode.cn id=968 lang=python3
#
# [968] 监控二叉树
#

# @lc code=start
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def traversal(self, root):
        if not root: return 2
        left = self.traversal(root.left)
        right = self.traversal(root.right)
        if left == 2 and right == 2:
            return 0
        elif left == 0 or right == 0:
            self.res += 1
            return 1
        else:
            return 2
    def minCameraCover(self, root: Optional[TreeNode]) -> int:
        self.res = 0
        if self.traversal(root) == 0:
            self.res += 1
        return self.res
# @lc code=end

File: 算法/代码随想录算法PDF/00调试源代码/968.监控二叉树.cpp
/*
 * @lc app=leetcode.cn id=968 lang=cpp
 *
 * [968] 监控二叉树
 */

// @lc code=start
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int res;
    int traversal(TreeNode* root) {
        if (!root) return 2;  // 终止条件
        int left = traversal(root->left);  // 后序递归
        int right = traversal(root->right);
        // 单层逻辑
        if (left == 2 && right == 2) return 0;
        if (left == 0 || right == 0) {
            res++;
            return 1;
        }
        if (left == 1 || right == 1) return 2;  // 这行必须在下面，优先保证监控
        return -1;  // 占位用，不会到这里
    }
    int minCameraCover(TreeNode* root) {
        res = 0;
        if (traversal(root) == 0) {
            res++;
        }
        return res;
    }
};
// @lc code=end

